endpoint_options.go

```diff
@@ -0,0 +1,310 @@
+type endpointOpts struct {
+	poolingEnabled          bool
+	bindings                []string
+	description             string
+	metadata                string
+	proxyProtocol           string
+	agentTLSConfig          *tls.Config
+	trafficPolicy           string
+	url                     string
+	upstreamProtocol        string
+	upstreamURL             string
+	upstreamTLSClientConfig *tls.Config
+}
+
+// defaultEndpointOpts returns the default options for an endpoint.
+func defaultEndpointOpts() *endpointOpts {
+	return &endpointOpts{}
+}
+
+// WithPoolingEnabled controls whether the endpoint supports connection pooling.
+//
+// See https://ngrok.com/docs/universal-gateway/endpoint-pooling/
+func WithPoolingEnabled(pool bool) EndpointOption {
+	return func(opts *endpointOpts) {
+		opts.poolingEnabled = pool
+	}
+}
+
+// WithBindings sets the endpoint's bindings.
+//
+// See https://ngrok.com/docs/universal-gateway/bindings/
+func WithBindings(bindings ...string) EndpointOption {
+	return func(opts *endpointOpts) {
+		opts.bindings = bindings
+	}
+}
+
+// WithDescription sets a human-readable description for the endpoint.
+func WithDescription(desc string) EndpointOption {
+	return func(opts *endpointOpts) {
+		opts.description = desc
+	}
+}
+
+// WithMetadata sets opaque, machine-readable metadata for the endpoint.
+func WithMetadata(meta string) EndpointOption {
+	return func(opts *endpointOpts) {
+		opts.metadata = meta
+	}
+}
+
+// WithProxyProtocol sets the proxy protocol for the endpoint. Valid values are
+// '1' or '2'.
+//
+// See https://ngrok.com/docs/agent/config/v3/#upstreamproxy_protocol
+func WithProxyProtocol(proxyProtocol string) EndpointOption {
+	return func(opts *endpointOpts) {
+		opts.proxyProtocol = proxyProtocol
+	}
+}
+
+// WithAgentTLSTermination sets a TLS configuration that the agent will use to
+// terminate connections received on the Endpoint.
+//
+// See https://ngrok.com/docs/agent/agent-tls-termination/
+func WithAgentTLSTermination(config *tls.Config) EndpointOption {
+	return func(opts *endpointOpts) {
+		opts.agentTLSConfig = config
+	}
+}
+
+// WithTrafficPolicy defines the Endpoint's Traffic Policy.
+//
+// See https://ngrok.com/docs/traffic-policy/
+func WithTrafficPolicy(policy string) EndpointOption {
+	return func(opts *endpointOpts) {
+		opts.trafficPolicy = policy
+	}
+}
+
+// WithURL defines the Endpoint's URL.
+func WithURL(urlSpec string) EndpointOption {
+	return func(opts *endpointOpts) {
+		opts.url = urlSpec
+	}
+}
+
+// configureEndpoint creates the appropriate tunnel configuration based on the URL
+// scheme and options
+func configureEndpoint(scheme string, endpointOpts *endpointOpts) (config.Tunnel, error) {
+	switch scheme {
+	case "http", "https":
+		return configureHTTPEndpoint(endpointOpts)
+	case "tcp":
+		return configureTCPEndpoint(endpointOpts)
+	case "tls":
+		return configureTLSEndpoint(endpointOpts)
+	default:
+		return nil, fmt.Errorf("unsupported URL scheme: %s", scheme)
```

Comment:
nit: It'd be nice to have the allowed options as enums or have the error message return the allowed strings so users trying it and getting it wrong know what is accepted more easily.

Reply:
good call, ty, will fix. once we do a little more surgery on this we'll be able to remove this entirely. this switch is really only here to deal with the legacy API